### ISSUE TITLE
Parallel-inconsistent Node locations on displaced meshes

### ISSUE THREAD

**Issue by roystgnr** [idaholab/moose#9626]
### Description of the enhancement or error report

Displaced mesh objects in MOOSE can (and typically do?) become inconsistent between processors.

### Rationale for the enhancement or information for reproducing the error

```METHOD=dbg ./run_tests -p 2 --distributed-mesh --recover --re materials/output.block_displaced```
is the first way to trigger the problem that I found.

MOOSE is only calling UpdateDisplacedMeshThread on its internal _active_semilocal_node_range.  So we displace all the nodes on local and MOOSE-ghosted elements, but we don't displace all ghosted nodes.  This is arguably *good* in the ReplicatedMesh case, where every node is ghosted but you only care about a fraction of them and you can get away with not wasting time displacing the rest.  In the DistributedMesh case, though, we have parallel mesh modification code that depends on node locations being consistent across processors, and we have a ton of debug code that asserts such consistency. results.

In the test case above, what happens is that a MeshSerializer attempts to serialize the mesh to processor 0 for Exodus output, the assertions in parallel_node.C discover that a node is in two different places on two different processors, and libMesh screams and dies.

### Identified impact
(i.e. Internal object changes, limited interface changes, public API change, or a list of specific applications impacted)

This came to my attention because it may be the last bug blocking #8410.

"Node locations are in sync between processors" is such a fundamental libMesh assumption that I can't imagine how many advanced features could be broken if that assumption isn't met.  As a start:  If we don't have correct node locations on neighboring elements then we will return the wrong results from flux based error estimators, as well as compute the wrong constraint equations on non-Lagrange elements.  If we don't have correct node locations on parent elements then we will compute the wrong constraint equations on Lagrange elements, and the wrong solution projection values on any elements.  If we don't have correct node locations on *arbitrarily distant layers* of elements then we will return the wrong results from patch recovery code.  If we don't have correct node locations on every ghosted element on a distributed mesh then we trigger assertions like the one above, and if we were to disable those assertions then several distributed mesh operations (including repartitioning and adaptive coarsening) would become race conditions for which failure could silently corrupt the mesh.

On the other hand, the obvious fix of "displace every node in the mesh" would turn an O(N_elem/N_proc) operation into an O(N_elem) operation in the ReplicatedMesh case.  I don't think this is significant (just creating a ReplicatedMesh in the first place is O(N_elem); if you don't like that then go DistributedMesh) but I'll wait for others to chime in before I go about pessimizing MOOSE code.

**Comment by permcody** [comment]
Yikes - first of all, nice description and yes this is a big problem. I don't have a strong opinion of how it's fixed as long as we get there. You talk about how we've managed some dumb luck about not displacing remote nodes in the ReplicatedMesh case, but is that really always the case? What about repartitioning if we turn on adaptivity? I don't think we do that very often in our mechanics problems, but it's not unheard of either. Some of the phase-field guys are running mechanics+phase_field and I believe they turn on adaptivity in those cases.

Unless it's a very simple optimization that we know works for ReplicatedMesh under all use cases, I don't see any problem with just displacing everything, that we we don't get left with any gotchas later.

**Comment by permcody** [comment]
Tag @idaholab/moose-team

**Comment by roystgnr** [comment]
The fix here isn't as simple as I thought it would be.

We can't use the provided solution vector to displace the mesh without a communication step first, because the mesh (if it's a ReplicatedMesh or if it's a DistributedMesh whose geometric ghosting exceeds its algebraic ghosting) will have nodes whose dof coefficients aren't stored in the default-ghosted vectors.

We can't just serialize the provided solution vector, because then it wouldn't scale with 1/N_proc in the DistributedMesh case either, which is unacceptable.  (and even in the ReplicatedMesh case, we want O(N_elem*N_dim), not O(N_elem*N_vars))

We'll need a second thread functor class, called from the constructor of the first, to do a sweep through all nodes and build a custom send_list with the displacement variables we need.  That won't be hard (I'll steal the libMesh BuildProjectionList code and simplify) but counting boilerplate it'll be a +400 line diff rather than a -4 line diff.

I still think this is the right fix, especially if @bwspenc has run into similar problems already, but if anyone has objections, speak now or once I've created the PR hold your peace.

**Comment by roystgnr** [comment]
And now that fix is all done, and seems to work fine for ReplicatedMesh...

But it's only solved the most superficial problem for DistributedMesh, because there we appear to be losing sync between the reference mesh and the displaced mesh.  The partitionings match (although in PartitionerAction.C it looks like our mechanism is just "use the same partitioner object for both"?), but there are elements which are being properly remoted on the reference mesh but *not* on the displaced mesh, which then causes the displaced mesh to be confused when it can't query the reference mesh to ask about displacements.

**Comment by roystgnr** [comment]
Okay, now I get it.

We're outputting Exodus in that test.  With use_displaced = true.  So the displaced mesh gets output, and because Exodus requires serialization we serialize it to processor 0.  For efficiency's sake we leave it serialized-to-0.  We don't ever serialize the reference mesh, so after the first output step they're out of sync, so in the next residual evaluation the update-displaced-mesh step fails.

**Comment by friedmud** [comment]
Coming in late...

This was an intentional part of the design.  No need to move nodes that the local processor doesn't care about (which means that many nodes become "inconsistent" - but we don't care).  If that means that we can't output the displaced mesh properly in parallel... then we should disable that.  No reason to do extra work and communication that isn't going to help any simulation.

**Comment by roystgnr** [comment]
True story: I tried to post the pissy rant below, and Github's web server chose that 60-second window to die on me and refuse to reload.  This looks suspiciously like some guardian angel granting me an opportunity to sound like less of a jackass, but I'm not sure I can actually do that, so as a compromise let me apologize in advance for sounding like a pissy jackass; it's not intended personally.

@friedmud:
> No reason to do extra work

I did acknowledge that:

> This is arguably good in the ReplicatedMesh case, where every node is ghosted but you only care about a fraction of them and you can get away with not wasting time displacing the rest.

@friedmud:
> which means that many nodes become "inconsistent" - but we don't care

> If that means that we can't output the displaced mesh properly in parallel...

If only that was the only problem, I'd just mark it WONTFIX and move on for now.  But:

> "Node locations are in sync between processors" is such a fundamental libMesh assumption that I can't imagine how many advanced features could be broken if that assumption isn't met. As a start: If we don't have correct node locations on neighboring elements then we will return the wrong results from flux based error estimators, as well as compute the wrong constraint equations on non-Lagrange elements. If we don't have correct node locations on parent elements then we will compute the wrong constraint equations on Lagrange elements, and the wrong solution projection values on any elements. If we don't have correct node locations on arbitrarily distant layers of elements then we will return the wrong results from patch recovery code. If we don't have correct node locations on every ghosted element on a distributed mesh then we trigger assertions like the one above, and if we were to disable those assertions then several distributed mesh operations (including repartitioning and adaptive coarsening) would become race conditions for which failure could silently corrupt the mesh.

We're finally starting to see displaced meshes giving correct solutions with AMR, fixing a years-old bug.  We're hopefully closer to getting them working with XFEM.  We still have hundreds of modules tests failing in parallel, not just distributed, but replicated in parallel, and debugging that is going to move from "hard" to "impossible" if I can't even assume that a corrupted mesh is corrupted.  At least for the failures I've been working on, I'm not talking about I/O, I'm talking about segmentation faults during execution and incorrect solutions.

Let's please postpone worry about how to "optimize" O(N_elem/N_proc*log N_elem + N_elem/N_proc) -> O(N_elem log N_elem) until *after* we handle the transition from broken -> working.  The old code was undefined behavior, the equivalent of removing a threading lock because you think it runs faster and you can run some cases without seeing any mistakes.  If you want a "HalfReplicatedMesh" that supports being replicated-but-not-really, we can do that, but it will take more time to get correct, will cost extra communication that will be more expensive than the full elements loop in many cases, and will certainly be less of an optimization than just getting MOOSE DistributedMesh support completely user-proofed.

**Comment by friedmud** [comment]
Ok - at least you gave a warning about how "pissy" it was 😉 

I wasn't necessarily advocating for the old behavior... I was just trying to shed some light not he fact that I knew I was doing that... and I was doing it "on purpose" (or "with some purpose").  It wasn't just an oversight or a blind bug.

Back when I made that decision and made that code work we were working at the VERY edge of what was possible and it was important not to fully synchronize the meshes.  But I'm ok with taking a step back from that and doing what's more reasonable these days instead.

**Comment by roystgnr** [comment]
Thanks!  I'm sorry, I didn't mean to imply that this wasn't a deliberate choice.

It might even be the case that "HalfReplicatedMesh" is so easy/cheap to accomplish that we'll just call it "ReplicatedMesh" and then re-disable the displacement of remote nodes on that, eventually.  (Although I suspect and hope that, by that time, it'll be cheaper and easier and better verified to just say "use DistributedMesh")  It's just that there's literally tens of thousands of lines of libMesh code that have been written with the assumption that a "half-replicated" node will never happen, and merely *verifying* that we accidentally made all/most of that code half-replicated-compatible would be a terrifying task.  If I'm going to start going line-by-line to root out atavistic assumptions, I'd rather use that on something productive like anisotropic adaptivity support.

**Comment by friedmud** [comment]
Sounds good
